src/pyromb/core/gis/builder.py

```diff
@@ -55,23 +58,43 @@ def basin(self, centroid: VectorLayer, basin: VectorLayer) -> list:
             A list of the basin objects.
         """
         basins = []
+        # Precompute Shapely polygons for all basins
+        basin_geometries = [basin.shapely_geometry(j) for j in range(len(basin))]
+
         for i in range(len(centroid)):
-            min = 0
-            d = 999
-            s = centroid.geometry(i)
+            centroid_geom = centroid.shapely_geometry(i)
```

Comment:
I used shapely as I'm more familiar with that. GDAL is annoying to install outside of QGIS... 
Shall see what the equivalent ogr2ogr commands are.

Reply:
GDAL can be installed with Conda without too much trouble. ogr2ogr is a command line tool is it not? this would require a subprocess to run which I would not want to do. 

The other, probably a more appropriate way, is to create a new operation under the math.geometry module called "contains" which given two geometries returns whether one contains the other. This contains functionality can be implemented by a geoprocessing library such as GDAL. but then can also be easily reimplemented by some other library (shapely) if needed. This would ensure the external geoprocessing library is only being used in one place allowing easy override. 

This is also relevant to geometry validation. There should be wrapped in the math.geometry module and then used the math.geometry module in the Builder class.

Reply:
Have got it working with osgeo through app.py, but shall check during the week what is required to make it work via QGIS. I assume one of the classes is going to get angry when dealing with the QGIS classes that need need to have a geometry method converted into something ogr likes.

Have tried to keep osgeo in geometry.py only, but I think it might have spread a bit further. Not sure of the best way to take it out of the validate_shapefile_geometries function.

My understanding is that GDAL is for the raster work, OGR is for the vector work. Or those are the tools I use on the command line anyway. They are also available as python and C bindings - this is the import I am using: from osgeo import ogr